cpp/src/arrow/array/data.h

```diff
@@ -232,6 +233,20 @@ struct ARROW_EXPORT ArrayData {
     return null_count.load() != length;
   }
 
+  Status AlignBuffers() {
+    // align buffers according to their data type's layout
+    for (auto&& [buffer, layout] : internal::Zip(buffers, type->layout().buffers)) {
+      if (layout.kind == DataTypeLayout::FIXED_WIDTH && buffer->address() % layout.byte_width) {
+        RETURN_NOT_OK(Buffer::Copy(buffer, buffer->memory_manager()).Value(&buffer));
+      }
+    }
+    // align children data recursively
+    for (unsigned int i=0; i<child_data.size(); i++) {
```

Comment:
nit: you could iterate with `for (auto& child : child_data)` and avoid the explicit index?

Reply:
much better!